examples/suit_update: use shell module

aabadie opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

This PR adds the shell module to the `suit_update` example. The nanocoap server is moved to its own thread. The Python test is adapted to add a call to ifconfig in order to retrieve the IPv6 address of the device.

Opening this PR was requested by @fjmolinas in #12471, which now depends on this one.

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure

Following the automatic testing procedure in the README should work.

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references

#12471 

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

miri64 commented on examples/suit_update/main.c:
```diff
@@ -54,24 +70,18 @@ int main(void)
     }
 
-    /* nanocoap_server uses gnrc sock which uses gnrc which needs a msg queue */
-    msg_init_queue(_main_msg_queue, MAIN_QUEUE_SIZE);
```
Only for commands that receive packets via GNRC like e.g. `ping6`.

miri64 replied:
The comment above is wrong though. `sock` doesn't require a message queue, as they bring their own.

fjmolinas replied:
> Only for commands that receive packets via GNRC like e.g. ping6.

Ok, that is why I'm getting the hardfault then. Thanks @miri64!

@aabadie  Could you add a `msg_queue` and a comment stating why?

aabadie commented:
I can reproduce your crash locally. It happens on the notify request. in #12471, it doesn't happen when triggering the update from the shell.
So there's something broken with the nanocoap server. Maybe the thread stack allocated for the server is not enough ? I'll try to find the problem tomorrow.

aabadie commented:
@fjmolinas, increasing the stack (e.g. use the same stack size than main) for the nanocoap server thread fixed the crash.
The automatic test is now passing: tested with nrf52840-mdk and samr21-xpro. I can show you IRL if you want.

fjmolinas commented on examples/suit_update/main.c:
```diff
@@ -19,20 +19,40 @@
 #define COAP_INBUF_SIZE (256U)
 
+/* Nanocoap server requires the same stack size as main */
+static char _nanocoap_server_stack[THREAD_STACKSIZE_MAIN];
```
Why does it require this size?

aabadie commented:
<details><summary>Here the partial test output for samr21-xpro:</summary>

```
$ make BOARD=samr21-xpro -C examples/suit_update test
make: Entering directory '/work/riot/RIOT/examples/suit_update'
make -C ethos
make[2]: Nothing to be done for 'all'.
make -C uhcpd
make[2]: Nothing to be done for 'all'.
make -C ethos
make[3]: Nothing to be done for 'all'.
make -C uhcpd
make[3]: Nothing to be done for 'all'.
/work/riot/RIOT/dist/tools/ethos/ethos riot0 /dev/ttyACM0
----> ethos: sending hello.
----> ethos: activating serial pass through.
----> ethos: hello reply received
----> ethos: hello received
gnrc_uhcpc: Using 5 as border interface and 0 as wireless interface.
gnrc_uhcpc: only one interface found, skipping setup.
main(): This is RIOT! (Version: 2020.01-devel-273-gee0d5-pr/examples/suit_example_shell)
RIOT SUIT update example application
running from slot 0
slot start addr = 0x1000
Image magic_number: 0x544f4952
Image Version: 0x5daf0b9d
Image start address: 0x00001100
Header chksum: 0xc6fc17ee

suit_coap: started.
ifconfig
Starting the shell
> ifconfig
Iface  5  HWaddr: 02:6B:3F:A9:DF:1E 
          L2-PDU:1500 MTU:1500  HL:64  RTR  
          RTR_ADV  
          Source address length: 6
          Link type: wired
          inet6 addr: fe80::6b:3fff:fea9:df1e  scope: link  TNT[1]
          inet6 addr: fe80::2  scope: link  VAL
pinging node...
PING fe80::2%riot0(fe80::2%riot0) 56 data bytes

--- fe80::2%riot0 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 38.976/38.976/38.976/0.000 ms
pinging node succeeded.
compiling /work/riot/RIOT/dist/tools/riotboot_gen_hdr/bin/genhdr...
make: Nothing to be done for 'all'.
creating /work/riot/RIOT/examples/suit_update/bin/samr21-xpro/suit_update-slot0.1571752862.riot.bin...
creating /work/riot/RIOT/examples/suit_update/bin/samr21-xpro/suit_update-slot1.1571752862.riot.bin...
/work/riot/RIOT/dist/tools/suit_v4/gen_manifest.py \
  --template /work/riot/RIOT/dist/tools/suit_v4/test-2img.json \
  --urlroot coap://[fd00:dead:beef::1]/fw/samr21-xpro \
  --seqnr 1571752862 \
  --uuid-vendor "riot-os.org" \
  --uuid-class samr21-xpro \
  --offsets 0x1000,133120 \
  -o /work/riot/RIOT/examples/suit_update/bin/samr21-xpro/suit_update-riot.suitv4.1571752862.bin \
  /work/riot/RIOT/examples/suit_update/bin/samr21-xpro/suit_update-slot0.1571752862.riot.bin /work/riot/RIOT/examples/suit_update/bin/samr21-xpro/suit_update-slot1.1571752862.riot.bin
{'structure-version': 1, 'sequence-number': 1571752862, 'components': [{'id': ['00']}], 'common': [{'condition-vendor-id': '547d0d746d3a5a9296624881afd9407b'}, {'condition-class-id': '8818989ea2575994ac9a554b77898083'}, {'directive-run-conditional': [{'directive-set-component': 0}, {'condition-component-offset': 4096}, {'directive-set-var': {'image-size': 86408, 'image-digest': [1, b'U\x9f\xff2\x93\xa9@kn5\x8b3FH\xff\t\xc7)z\x9b\xc5\xff\x1b\n4)\xa3\x0f\xcf\xba\xf7\xe9']}}]}, {'directive-run-conditional': [{'directive-set-component': 0}, {'condition-component-offset': 133120}, {'directive-set-var': {'image-size': 86408, 'image-digest': [1, b'q\x8c3z\xd9\x8a\x13N\xf0\xf0]\xa0\xda\xab\xa5\xf1\xc9]i\xbb\xb9`\xb1\xfd]j\x91\x00\xf7\xa5\xd0\x9f']}}]}], 'apply-image': [{'directive-run-conditional': [{'directive-set-component': 0}, {'condition-component-offset': 4096}, {'directive-set-var': {'uris': [[0, 'coap://[fd00:dead:beef::1]/fw/samr21-xpro/suit_update-slot0.1571752862.riot.bin']]}}]}, {'directive-run-conditional': [{'directive-set-component': 0}, {'condition-component-offset': 133120}, {'directive-set-var': {'uris': [[0, 'coap://[fd00:dead:beef::1]/fw/samr21-xpro/suit_update-slot1.1571752862.riot.bin']]}}]}, {'directive-set-component': True}, {'directive-fetch': None}], 'system-verification': [{'directive-set-component': True}, {'condition-image': None}], 'run-image': [{'directive-set-component': 0}, {'directive-run': None}]}
/work/riot/RIOT/dist/tools/suit_v4/sign-04.py \
  /work/riot/RIOT/keys/default /work/riot/RIOT/keys/default.pub /work/riot/RIOT/examples/suit_update/bin/samr21-xpro/suit_update-riot.suitv4.1571752862.bin /work/riot/RIOT/examples/suit_update/bin/samr21-xpro/suit_update-riot.suitv4_signed.1571752862.bin
None
published "/work/riot/RIOT/examples/suit_update/bin/samr21-xpro/suit_update-riot.suitv4.1571752862.bin"
       as "coap://[fd00:dead:beef::1]/fw/samr21-xpro/suit_update-riot.suitv4.1571752862.bin"
published "/work/riot/RIOT/examples/suit_update/bin/samr21-xpro/suit_update-riot.suitv4.latest.bin"
       as "coap://[fd00:dead:beef::1]/fw/samr21-xpro/suit_update-riot.suitv4.latest.bin"
published "/work/riot/RIOT/examples/suit_update/bin/samr21-xpro/suit_update-riot.suitv4_signed.1571752862.bin"
       as "coap://[fd00:dead:beef::1]/fw/samr21-xpro/suit_update-riot.suitv4_signed.1571752862.bin"
published "/work/riot/RIOT/examples/suit_update/bin/samr21-xpro/suit_update-riot.suitv4_signed.latest.bin"
       as "coap://[fd00:dead:beef::1]/fw/samr21-xpro/suit_update-riot.suitv4_signed.latest.bin"
published "/work/riot/RIOT/examples/suit_update/bin/samr21-xpro/suit_update-slot0.1571752862.riot.bin"
       as "coap://[fd00:dead:beef::1]/fw/samr21-xpro/suit_update-slot0.1571752862.riot.bin"
published "/work/riot/RIOT/examples/suit_update/bin/samr21-xpro/suit_update-slot1.1571752862.riot.bin"
       as "coap://[fd00:dead:beef::1]/fw/samr21-xpro/suit_update-slot1.1571752862.riot.bin"
aiocoap-client -m POST "coap://[fe80::2%riot0]/suit/trigger" \
	--payload "coap://[fd00:dead:beef::1]/fw/samr21-xpro/$(basename /work/riot/RIOT/examples/suit_update/bin/samr21-xpro/suit_update-riot.suitv4_signed.1571752862.bin)" && \
	echo "Triggered [fe80::2%riot0] to update."
Triggered [fe80::2%riot0] to update.
          inet6 group: ff02::2
          inet6 group: ff02::1
          inet6 group: ff02::1:ffa9:df1e
          inet6 group: ff02::1:ff00:2
          
> suit: received URL: "coap://[fd00:dead:beef::1]/fw/samr21-xpro/suit_update-riot.suitv4_signed.1571752862.bin"
suit_coap: trigger received
suit_coap: downloading "coap://[fd00:dead:beef::1]/fw/samr21-xpro/suit_update-riot.suitv4_signed.1571752862.bin"
suit_coap: got manifest with size 545
suit: verifying manifest signature...
suit: validated manifest version
)riotboot_hdr_validate: riotboot_hdr magic number invalid
suit: validated sequence number
)validating vendor ID
Comparing 547d0d74-6d3a-5a92-9662-4881afd9407b to 547d0d74-6d3a-5a92-9662-4881afd9407b from manifest
validating vendor ID: OK
validating class id
Comparing 8818989e-a257-5994-ac9a-554b77898083 to 8818989e-a257-5994-ac9a-554b77898083 from manifest
validating class id: OK
Comparing manifest offset 4096 with other slot offset 133120
Comparing manifest offset 133120 with other slot offset 133120
Comparing manifest offset 4096 with other slot offset 133120
Comparing manifest offset 133120 with other slot offset 133120
riotboot_flashwrite: initializing update to target slot 1
riotboot_flashwrite: processing bytes 4-63
riotboot_flashwrite: processing bytes 64-127
riotboot_flashwrite: processing bytes 128-191
riotboot_flashwrite: processing bytes 192-255
riotboot_flashwrite: processing bytes 256-319
riotboot_flashwrite: processing bytes 320-383
[...]
riotboot_flashwrite: processing bytes 85952-86015
riotboot_flashwrite: processing bytes 86016-86079
riotboot_flashwrite: processing bytes 86080-86143
riotboot_flashwrite: processing bytes 86144-86207
riotboot_flashwrite: processing bytes 86208-86271
riotboot_flashwrite: processing bytes 86272-86335
riotboot_flashwrite: processing bytes 86336-86399
riotboot_flashwrite: processing bytes 86400-86407
riotboot: verifying digest at 0x20003bb7 (img at: 0x20800 size: 86408)
suit_v4_parse() success
SUIT policy check OK.
suit_coap: finalizing image flash
riotboot_flashwrite: riotboot flashing completed successfully
Image magic_number: 0x544f4952
Image Version: 0x5daf0b9e
Image start address: 0x00020900
Header chksum: 0xb7020ff1

----> ethos: hello received
Failed to send flush request: Operation not permitted
gnrc_uhcpc: Using 5 as border interface and 0 as wireless interface.
gnrc_uhcpc: only one interface found, skipping setup.
main(): This is RIOT! (Version: 2020.01-devel-273-gee0d5-pr/examples/suit_example_shell)
RIOT SUIT update example application
running from slot 1
slot start addr = 0x20800
Image magic_number: 0x544f4952
Image Version: 0x5daf0b9e
Image start address: 0x00020900
Header chksum: 0xb7020ff1

suit_coap: started.
ifconfig
Starting the shell
> ifconfig
Iface  5  HWaddr: 02:6B:3F:A9:DF:1E 
          L2-PDU:1500 MTU:1500  HL:64  RTR  
          RTR_ADV  
          Source address length: 6
          Link type: wired
          inet6 addr: fe80::6b:3fff:fea9:df1e  scope: link  TNT[1]
          inet6 addr: fe80::2  scope: link  VAL
pinging node...
PING fe80::2%riot0(fe80::2%riot0) 56 data bytes

--- fe80::2%riot0 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 38.628/38.628/38.628/0.000 ms
pinging node succeeded.
compiling /work/riot/RIOT/dist/tools/riotboot_gen_hdr/bin/genhdr...
make: Nothing to be done for 'all'.
creating /work/riot/RIOT/examples/suit_update/bin/samr21-xpro/suit_update-slot0.1571752863.riot.bin...
creating /work/riot/RIOT/examples/suit_update/bin/samr21-xpro/suit_update-slot1.1571752863.riot.bin...
/work/riot/RIOT/dist/tools/suit_v4/gen_manifest.py \
  --template /work/riot/RIOT/dist/tools/suit_v4/test-2img.json \
  --urlroot coap://[fd00:dead:beef::1]/fw/samr21-xpro \
  --seqnr 1571752863 \
  --uuid-vendor "riot-os.org" \
  --uuid-class samr21-xpro \
  --offsets 0x1000,133120 \
  -o /work/riot/RIOT/examples/suit_update/bin/samr21-xpro/suit_update-riot.suitv4.1571752863.bin \
  /work/riot/RIOT/examples/suit_update/bin/samr21-xpro/suit_update-slot0.1571752863.riot.bin /work/riot/RIOT/examples/suit_update/bin/samr21-xpro/suit_update-slot1.1571752863.riot.bin
{'structure-version': 1, 'sequence-number': 1571752863, 'components': [{'id': ['00']}], 'common': [{'condition-vendor-id': '547d0d746d3a5a9296624881afd9407b'}, {'condition-class-id': '8818989ea2575994ac9a554b77898083'}, {'directive-run-conditional': [{'directive-set-component': 0}, {'condition-component-offset': 4096}, {'directive-set-var': {'image-size': 86408, 'image-digest': [1, b'K\x1f\xad\x12\xb5\x9b\xca\x80\xb6\x1f\xe4,\x9c\xf1\x9f\xdfF\x03}v\xd3\xd7\xaa\x96\x9c\x1c0,\x00\xdd6\x8c']}}]}, {'directive-run-conditional': [{'directive-set-component': 0}, {'condition-component-offset': 133120}, {'directive-set-var': {'image-size': 86408, 'image-digest': [1, b'XU\xcam\xc8\xe4!|Ci\xa4\xc0\xd8\x0bN>\x00\xdb\xb4\xc0\xb1\xdf\xd4\xda\x81\xf3+`6\xe75\x90']}}]}], 'apply-image': [{'directive-run-conditional': [{'directive-set-component': 0}, {'condition-component-offset': 4096}, {'directive-set-var': {'uris': [[0, 'coap://[fd00:dead:beef::1]/fw/samr21-xpro/suit_update-slot0.1571752863.riot.bin']]}}]}, {'directive-run-conditional': [{'directive-set-component': 0}, {'condition-component-offset': 133120}, {'directive-set-var': {'uris': [[0, 'coap://[fd00:dead:beef::1]/fw/samr21-xpro/suit_update-slot1.1571752863.riot.bin']]}}]}, {'directive-set-component': True}, {'directive-fetch': None}], 'system-verification': [{'directive-set-component': True}, {'condition-image': None}], 'run-image': [{'directive-set-component': 0}, {'directive-run': None}]}
/work/riot/RIOT/dist/tools/suit_v4/sign-04.py \
  /work/riot/RIOT/keys/default /work/riot/RIOT/keys/default.pub /work/riot/RIOT/examples/suit_update/bin/samr21-xpro/suit_update-riot.suitv4.1571752863.bin /work/riot/RIOT/examples/suit_update/bin/samr21-xpro/suit_update-riot.suitv4_signed.1571752863.bin
None
published "/work/riot/RIOT/examples/suit_update/bin/samr21-xpro/suit_update-riot.suitv4.1571752863.bin"
       as "coap://[fd00:dead:beef::1]/fw/samr21-xpro/suit_update-riot.suitv4.1571752863.bin"
published "/work/riot/RIOT/examples/suit_update/bin/samr21-xpro/suit_update-riot.suitv4.latest.bin"
       as "coap://[fd00:dead:beef::1]/fw/samr21-xpro/suit_update-riot.suitv4.latest.bin"
published "/work/riot/RIOT/examples/suit_update/bin/samr21-xpro/suit_update-riot.suitv4_signed.1571752863.bin"
       as "coap://[fd00:dead:beef::1]/fw/samr21-xpro/suit_update-riot.suitv4_signed.1571752863.bin"
published "/work/riot/RIOT/examples/suit_update/bin/samr21-xpro/suit_update-riot.suitv4_signed.latest.bin"
       as "coap://[fd00:dead:beef::1]/fw/samr21-xpro/suit_update-riot.suitv4_signed.latest.bin"
published "/work/riot/RIOT/examples/suit_update/bin/samr21-xpro/suit_update-slot0.1571752863.riot.bin"
       as "coap://[fd00:dead:beef::1]/fw/samr21-xpro/suit_update-slot0.1571752863.riot.bin"
published "/work/riot/RIOT/examples/suit_update/bin/samr21-xpro/suit_update-slot1.1571752863.riot.bin"
       as "coap://[fd00:dead:beef::1]/fw/samr21-xpro/suit_update-slot1.1571752863.riot.bin"
aiocoap-client -m POST "coap://[fe80::2%riot0]/suit/trigger" \
	--payload "coap://[fd00:dead:beef::1]/fw/samr21-xpro/$(basename /work/riot/RIOT/examples/suit_update/bin/samr21-xpro/suit_update-riot.suitv4_signed.1571752863.bin)" && \
	echo "Triggered [fe80::2%riot0] to update."
Triggered [fe80::2%riot0] to update.
          inet6 group: ff02::2
          inet6 group: ff02::1
          inet6 group: ff02::1:ffa9:df1e
          inet6 group: ff02::1:ff00:2
          
> suit: received URL: "coap://[fd00:dead:beef::1]/fw/samr21-xpro/suit_update-riot.suitv4_signed.1571752863.bin"
suit_coap: trigger received
suit_coap: downloading "coap://[fd00:dead:beef::1]/fw/samr21-xpro/suit_update-riot.suitv4_signed.1571752863.bin"
suit_coap: got manifest with size 545
suit: verifying manifest signature...
suit: validated manifest version
)suit: validated sequence number
)validating vendor ID
Comparing 547d0d74-6d3a-5a92-9662-4881afd9407b to 547d0d74-6d3a-5a92-9662-4881afd9407b from manifest
validating vendor ID: OK
validating class id
Comparing 8818989e-a257-5994-ac9a-554b77898083 to 8818989e-a257-5994-ac9a-554b77898083 from manifest
validating class id: OK
Comparing manifest offset 4096 with other slot offset 4096
Comparing manifest offset 133120 with other slot offset 4096
Comparing manifest offset 4096 with other slot offset 4096
Comparing manifest offset 133120 with other slot offset 4096
riotboot_flashwrite: initializing update to target slot 0
riotboot_flashwrite: processing bytes 4-63
riotboot_flashwrite: processing bytes 64-127
riotboot_flashwrite: processing bytes 128-191
riotboot_flashwrite: processing bytes 192-255
riotboot_flashwrite: processing bytes 256-319
riotboot_flashwrite: processing bytes 320-383
riotboot_flashwrite: processing bytes 384-447
riotboot_flashwrite: processing bytes 448-511
riotboot_flashwrite: processing bytes 512-575

[...]
riotboot_flashwrite: processing bytes 86144-86207
riotboot_flashwrite: processing bytes 86208-86271
riotboot_flashwrite: processing bytes 86272-86335
riotboot_flashwrite: processing bytes 86336-86399
riotboot_flashwrite: processing bytes 86400-86407
riotboot: verifying digest at 0x20003b78 (img at: 0x1000 size: 86408)
suit_v4_parse() success
SUIT policy check OK.
suit_coap: finalizing image flash
riotboot_flashwrite: riotboot flashing completed successfully
Image magic_number: 0x544f4952
Image Version: 0x5daf0b9f
Image start address: 0x00001100
Header chksum: 0xc70417f0

----> ethos: hello received
Failed to send flush request: Operation not permitted
gnrc_uhcpc: Using 5 as border interface and 0 as wireless interface.
gnrc_uhcpc: only one interface found, skipping setup.
main(): This is RIOT! (Version: 2020.01-devel-273-gee0d5-pr/examples/suit_example_shell)
RIOT SUIT update example application
running from slot 0
TEST PASSED

make: Leaving directory '/work/riot/RIOT/examples/suit_update'
```

</details>

fjmolinas commented:
I still don't understand the reason for the stack size, but I don't think it needs to be a blocking reason right now. I'll need to re-test.

fjmolinas commented on examples/suit_update/main.c:
```diff
@@ -54,24 +74,22 @@ int main(void)
-    msg_init_queue(_main_msg_queue, MAIN_QUEUE_SIZE);
-
-    puts("Waiting for address autoconfiguration...");
-    xtimer_sleep(3);
```
There is an issue with removing this, without giving time for address to configure the first time `ifconfig` is called the address is tentative `TNT[1]`, this causes the wrong address to be used (not the EUI64) address which the node uses to respond to COAP requests.

So keep `xtimer_sleep` or change the matching regex (since ping will make sure the node becomes available).

```
> ifconfig
Iface  5  HWaddr: 00:53:CB:6B:A0:79 
          L2-PDU:1500 MTU:1500  HL:64  RTR  
          RTR_ADV  
          Source address length: 6
          Link type: wired
          inet6 addr: fe80::253:cbff:fe6b:a079  scope: link  TNT[1]
          inet6 addr: fe80::2  scope: link  VAL
pinging node...
PING fe80::2%riot0(fe80::2%riot0) 56 data bytes
```

fjmolinas replied:
or add a sleep(1) in the test script.

aabadie replied:
Fixed

fjmolinas commented on examples/suit_update/tests/01-run.py:
```diff
@@ -121,13 +123,15 @@ def testfunc(child):
     for version in [current_app_ver + 1, current_app_ver + 2]:
+        child.expect_exact("suit_coap: started.")
+        child.expect_exact("Starting the shell")
+        time.sleep(1) # give some time for the network interface to be configured
```
```suggestion
        # give some time for the network interface to be configured
        time.sleep(1)
```

fjmolinas commented:
@aabadie some unrelated changes were included in your rebase, please fix and squash directly.

aabadie commented:
> some unrelated changes were included in your rebase, please fix and squash directly.

Sorry, I messed up with local changes... This is fixed now, squashed and rebased.
